### ISSUE TITLE
`pep484-style-positional-only-parameter` (`PYI063`) - false negative on `__new__` method

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#16410]
### Summary

```py
class Foo:
    def foo(self, __asdf: int): ... # error (correct)

    def __new__(self, __asdf: int): ... # no error (incorrect)
```
https://play.ruff.rs/52ad7e69-3721-4a1a-8ec4-fa7fa99f7282

### Version

0.9.7

**Comment by dhruvmanila** [comment]
It also seems like the rule only checks the first parameter: https://play.ruff.rs/8982da12-d242-4794-805c-fa13bc6531f0 which seems like a bug as well.

**Comment by Daverball** [comment]
> It also seems like the rule only checks the first parameter: https://play.ruff.rs/8982da12-d242-4794-805c-fa13bc6531f0 which seems like a bug as well.

The old special casing only worked if the first parameter was using double underscores, so any double underscores in a different position would more likely be a false positive. So I'm not convinced the other parameters should be flagged.

But it might make sense to flag all the directly following parameters with double underscores if the first one contains double underscores.

If agree that if we ever wanted to add an unsafe fix for this rule you'd definitely want it to rewrite `def (__a: int, __b: int)` to `def(a: int, b: int, /)`, rather than `def(a: int, \, __b: int)`.

**Comment by DetachHead** [comment]
> any double underscores in a different position would more likely be a false positive. So I'm not convinced the other parameters should be flagged.

why's that?

**Comment by Daverball** [comment]
> why's that?

Because it wouldn't have been recognized as a positional-only parameter by type checkers, so why should we tell people to turn it into one?

Now of course people can make mistakes and may have intended a parameter that isn't in the first position to be positional-only and just forgot to mark the previous parameter, but it seems just as likely to me that the parameter did indeed have double underscores in the original source code. I have seen it more than once when contributing stubs.

You always have to balance false negatives against false positives. Is it worth it to catch a few more things only to incorrectly flag others? Maybe, maybe not. I think the rule currently strikes a good balance, apart from the false negative with `__new__` you originally reported.

**Comment by DetachHead** [comment]
oh right. i didn't realize this is considered invalid:

```py
def foo(a: int, __b: str): ...
```
i just assumed it meant both `a` and `__b` are considered positional-only. [pyright reports an error](https://basedpyright.com/?typeCheckingMode=all&reportUnusedParameter=false&code=CYUwZgBGD20BQEMBcECWA7ALgGggfTwCMUBnTAJwEoUA6OoA) if you do this but [mypy doesn't](https://mypy-play.net/?mypy=latest&python=3.13&flags=show-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=379b0feb78c35d25776dbf9e544f46eb)

**Comment by joshuavetos** [comment]
I traced this to the NewMethod change in #13305. I also noticed PLE1519 and PLW0245 seem to have lost __new__ diagnostics from the same change. Should a fix here only cover PYI063, or should those related cases be included too?

**Comment by hxperl** [comment]
Hi — is anyone working on this?

@joshuavetos I saw you traced it to #13305 last week. Are you planning to open a PR? If not, I'd like to try it.

I don't want to duplicate your work, so I'll wait a bit before starting.

**Comment by joshuavetos** [comment]
Yeah, I’m planning to work on it. I was waiting on clarification about whether the related PLE1519/PLW0245 regressions should be included, but I can start with PYI063 and keep the scope narrow unless a maintainer says otherwise.

**Comment by hxperl** [comment]
Sounds good — all yours! Thanks for the heads up.

For what it's worth, I did confirm PLE1519 and PLW0245 lost `__new__` the same way, so your read on #13305 looks right.
